tensorflow/core/grappler/optimizers/auto_mixed_precision.cc

```diff
@@ -0,0 +1,1762 @@
+      const auto& arg_inds = output_arg_inds[i];
+      const OpDef::ArgDef& arg_def = op_def.output_arg(arg_inds.first);
+      TypeAttrId type_attr = GetTypeAttrId(arg_def, arg_inds.second);
+      if (!type_attr.attr_name.empty() &&
+          !node.attr().count(type_attr.attr_name)) {
+        return errors::InvalidArgument("Type attribute ", type_attr.attr_name,
+                                       " is not present in node ", node.name());
+      }
+      type2io_entry[type_attr].second.insert(i);
+      io2type_entry.second.push_back(type_attr);
+    }
+
+    // Also ensure that type attributes that aren't associated with any inputs
+    // or outputs (e.g., StackV2's elem_type) are added to the map.
+    for (const auto& attr : node.attr()) {
+      const string& attr_name = attr.first;
+      if (!attr_name.empty() && attr_name[0] == '_') continue;
+      const AttrValue& attr_value = attr.second;
+      const OpDef::AttrDef* attr_def = FindAttr(attr_name, op_def);
+      if (!attr_def) {
+        return errors::InvalidArgument("AttrDef not found for attribute ",
+                                       attr_name, " of node ", node.name());
+      }
+      if (attr_def->type() == "type") {
+        type2io_entry[TypeAttrId(attr_name)];
+      } else if (attr_def->type() == "list(type)") {
+        for (int i = 0; i < (int)attr_value.list().type_size(); ++i) {
+          type2io_entry[TypeAttrId(attr_name, i)];
+        }
+      }
+    }
+    return Status::OK();
+  }
+
+  // WARN: `graph_` must outlive this object (node pointers must remain valid).
+  const GraphDef* graph_ = nullptr;  // do not own
+
+  typedef absl::flat_hash_set<int> IntSet;
+  // Maps a type attr id -> (input port set, output port set)
+  typedef absl::flat_hash_map<TypeAttrId, std::pair<IntSet, IntSet>> Type2IOMap;
+  // Maps a node -> type attr mapping
+  absl::flat_hash_map<const NodeDef*, Type2IOMap> type2io_;
+  // Maps a port -> type attr id
+  typedef std::vector<TypeAttrId> TypeAttrIdVec;
+  // Maps a node -> (input port mapping, output port mapping)
+  absl::flat_hash_map<const NodeDef*, std::pair<TypeAttrIdVec, TypeAttrIdVec>>
+      io2type_;
+};
+
+struct NodeTypeId {
+  NodeTypeId(const NodeDef* _node, const TypeAttrId& _type_attr)
+      : node(_node), type_attr(_type_attr) {}
+
+  const NodeDef* node;
+  TypeAttrId type_attr;
+
+  bool operator==(const NodeTypeId& other) const {
+    return node == other.node && type_attr == other.type_attr;
+  }
+
+  template <typename H>
+  friend H AbslHashValue(H h, const NodeTypeId& nt) {
+    return H::combine(std::move(h), nt.node, nt.type_attr);
+  }
+};
+
+struct NodeTypeIdEdge {
+  NodeTypeIdEdge(const NodeTypeId& _src, const NodeTypeId& _dst)
+      : src(_src), dst(_dst) {}
+  NodeTypeId src;
+  NodeTypeId dst;
+};
+
+// This is just like GraphTopologyView but with (NodeDef, TypeAttrId) pairs as
+// the vertices instead of just NodeDef.
+class GraphTypeTopologyView {
```

Comment:
I think using a GraphTopologyView would suffice. I think including the types in the graph could hide certain relevant connections.

E.g., if we have an op with two inputs of type T and U respectively, and two outputs of type T and U. With GraphTypeTopologyView, the op would be represented by two unconnected nodes. But, U could affect the numerical stability of the T output, so there perhaps should be a connection.

In practice, there is no such op. But in practice, I also think a GraphTopologyView would behave similarly to a GraphTypeTopologyView and not impact correctness or performance. Duplicating code for a GraphTypeTopologyView is not great.

For now, the current approach is fine, bu we can discuss about changing this after this PR is submitted.

Reply:
I've added a TODO.

Reply:
I would prefer that we do not check all this duplicated code that is to some extent copy-and-pasted from existing Grappler code.

Reply:
While I agree the duplicated code is not great, I think we should leave it as a TODO and try changing this in a future PR. Reducing this duplication will result in fundamentally changing the algorithm used to color nodes and insert casts, and this new hypothetical algorithm is currently untested and may not work. Also, we are trying to get this PR in for the TF 1.14 branch, so that users can start using mixed precision as soon as possible.

Reply:
Sounds good to me.